5주차 과제

jangsc0000 opened the pull request:
잘부탁드립니다!

이번 주차는 

- 엔티티 -> 레포지토리-> 서비스 -> 컨트롤러,
- 자바독 -> 테스트 -> 클래스 구현 -> 수정 

순으로 개발해보겠습니다!

jangsc0000 committed:
[테스트] User 테스트 추가
- UserBuilder를 이용한 User 생성 테스트 및 사용 예시
- Setter를 이용한 User의 속성 변경 테스트 및 사용 예시

jangsc0000 committed:
[추가] User Entity 구현 추가

jangsc0000 committed:
[추가] UserRepository interface 추가

jangsc0000 committed:
[추가] UserService의 JavaDoc 추가

jangsc0000 committed:
[테스트] UserService의 테스트 추가
- Happy path의 컨텍스트 추가

jangsc0000 committed:
[추가] UserService 구현 추가
- id 조회 실패시, null 반환하도록 구현
- 'User를 찾을수 없다'는예외를 리턴하도록 변경 필요

jangsc0000 committed:
[테스트] UserService 미등록 유저 컨텍스트의 테스트
- findUserById, updateUser, deleteUser에 '등록되지않은 유저의 id가
  주어진다면' 컨텍스트 추가

jangsc0000 committed:
[추가] UserNotFound 예외처리 추가

jangsc0000 committed:
[리팩토링] UserServiceTest 코드 중복 제거

jangsc0000 committed:
[추가] UserController JavaDoc 추가

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,86 @@
+@Service
+public class UserService {
+
+    private UserRepository userRepository;
```
일단 할당되면 누군가가 바꾸는 실수를 할 수 없도록 이렇게 `final`을 붙여줍시다.
```suggestion
    private final UserRepository userRepository;
```

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,86 @@
+     *
+     * @param id 유저 id
+     * @return 주어진 id와 일치하는 유저
+     * @throws '유저를 찾지 못했다'는 예외
```
`@throws`에는 예외 클래스 이름도 적어주셔야 해요. JavaDoc이 빌드되면 코드 본문은 보이지 않거든요. 여기에 지정된 예외 클래스 이름은 IDE의 리팩토링 도구를 사용하면 클래스 이름을 바꿀 때 같이 수정됩니다.

```suggestion
     * @throws UserNotFoundException '유저를 찾지 못했다'는 예외
```

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,86 @@
+        User user = userRepository.findById(id).orElseThrow(() ->
+                new UserNotFoundException("유저 id: " + id + "를 찾을 수 없어, 업데이트할 수 없습니다."
+                ));
+        user.setName(source.getName());
+        user.setEmail(source.getEmail());
+        user.setPassword(source.getPassword());
+
+        return userRepository.save(user);
```
잘 작업하셨습니다. 난이도를 조금 더 높여서 작업해도 괜찮을 것 같아요. 시간적인 여유가 있다면 도전해 보세요. `setter`를 사용하지 않고, 생성자를 통해 업데이트된 내용을 가진 `User`를 생성하고 `userRepository`에 저장하도록 해보는 방법입니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,230 @@
+            return userService.findAllUsers();
+        }
+
+        @Nested
+        @DisplayName("등록된 유저가 있다면")
+        class Context_with_existed_user {
+            @BeforeEach
+            void prepare() {
+                prepareExitedUser();
+            }
```
너무너무 잘 하고 계세요. 그러니 다음 단계로도 더 올라가 봅시다!

```suggestion
        @Nested
        @DisplayName("등록된 유저가 한 명 있다면")
        class Context_with_existed_user {
            @BeforeEach
            void prepare() {
                prepareExitedUser();
            }
```

이렇게 바꾸고, 다음 컨텍스트들도 추가해 보세요.

- `등록된 유저가 2명 이상이라면`
- `등록된 유저가 아무도 없다면`

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,230 @@
+            @Test
+            @DisplayName("'유저를 찾을수 없다'는 예외를 던진다.")
+            void it_throws_not_found_user_exception() {
+                assertThatThrownBy(() -> subject(NOT_EXISTED_USER_ID, updateUser)).isInstanceOf(UserNotFoundException.class);
```
한 줄이 너무 길어요. 적절히 개행해서 리뷰하는 사람(동료들)이 읽기 좋게 해주세요.

johngrib commented on app/src/test/java/com/codesoom/assignment/domain/UserTest.java:
```diff
@@ -0,0 +1,62 @@
+@DisplayName("User")
+public class UserTest {
+    @Nested
+    @DisplayName("User 빌더는")
+    class Describe_UserBuilder{
+        @Test
+        @DisplayName("User를 생성한다.")
+        void it_returns_user(){
+            User user = User.builder()
+                            .id(1L)
+                            .email("honggd@gmail.com")
+                            .name("홍길동")
+                            .password("password")
+                            .build();
+
+            assertThat(user.getId()).isEqualTo(1L);
+            assertThat(user.getEmail()).isEqualTo("honggd@gmail.com");
+            assertThat(user.getName()).isEqualTo("홍길동");
+            assertThat(user.getPassword()).isEqualTo("password");
+        }
```
이 테스트는 불필요해 보여요.

johngrib reviewed (commented):
화이팅! 너무 잘 하고 계십니다!

jangsc0000 committed:
[테스트] UserController 테스트 추가
- Happy path의 컨텍스트만 추가

jangsc0000 committed:
[수정] UserService의 메소드 public으로 변경
- UserService의 메소드들이 private으로 구현되어있었음

jangsc0000 committed:
[추가] UserController 구현 추가
- create, update, delete 구현
- 테스트의 한글 에러 수정

jangsc0000 committed:
[리팩토링] UserService 수정
- updateUser에서 Setter 쓰지않고  Builder 사용하여 구현
- JavaDoc @thorws에 클래스명 추가
- userRepository 멤버에 final 추가

jangsc0000 committed:
[테스트] UserController 테스트에 컨텍스트 추가
- 등록되지 않은 유저 id에 대한 컨텍스트 추가

jangsc0000 committed:
[추가] User Not Found에 대한 ExceptionHandler 추가

jangsc0000 committed:
[리팩토링] UserData DTO 추가
- UserData DTO 추가
- UserData에 Validation 추가

jangsc0000 committed:
[추가] dozerMapper dependency 추가

johngrib commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,15 +1,23 @@
 
-    public static void main(String[] args) {
-        SpringApplication.run(App.class, args);
+    @Bean
```
`@Bean`의 기능과 역할에 대해 조사해서 설명하는 글을 작성해 보세요. 이 애노테이션은 아주 중요하니 이번 기회에 확실히 학습해 봅시다!

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,96 @@
+    private final UserRepository userRepository;
+    private final Mapper mapper;
+
+    @Autowired
+    public UserService(UserRepository userRepository, Mapper dozerMapper) {
+        this.userRepository = userRepository;
+        this.mapper = dozerMapper;
+    }
```
여기에서 `@Autowired`가 없어도 Spring은 `UserRepository`와 `Mapper`를 잘 주입해 줄 겁니다. 왜 그런지 이유를 조사해 보세요.
```suggestion
    public UserService(UserRepository userRepository, Mapper dozerMapper) {
        this.userRepository = userRepository;
        this.mapper = dozerMapper;
    }
```

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,96 @@
+     * @return 주어진 id와 일치하는 유저
+     * @throws UserNotFoundException '유저를 찾지 못했다'는 예외
+     */
+    public User findUserById(Long id) {
+        return userRepository.findById(id).orElseThrow(() ->
+                new UserNotFoundException("유저 id: " + id + "를 찾을 수 없어, 조회할 수 없습니다."));
+    }
```
`Long id`를 쓰고 있으니 `id`에 `null`이 들어올 가능성도 있어요. `id`가 `null`이라면 `"유저 id: " + id + "를 찾을 수 없어, 조회할 수 없습니다."`에서 문자열을 `+` 하다가 NPE가 던져질 겁니다. 이런 문제를 예방하기 위해 `String.format`을 사용해서 에러 메시지를 만드는 방법으로 바꿔 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/User.java:
```diff
@@ -0,0 +1,32 @@
+
+@Entity
+@Getter
+@Setter
```
만약 `@Setter`를 사용하지 않기로 동료들과 약속을 하게 됐다면 어떻게 해야 할 지 한 번 고민해 보세요.

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -0,0 +1,225 @@
+            @Test
+            @DisplayName("Not Found를 응답한다.")
+            void it_responses_not_found() throws Exception {
+                mockMvc.perform(delete("/users/0"))
```
여기에 있는 `0`이 유저의 아이디죠? 그렇다면 아이디를 분리해서 바로 위에 있는 컨텍스트 영역에서 선언해 주세요.

johngrib reviewed (commented):
잘 하고 계세요. 조금만 더 힘을 내봅시다.

jangsc0000 committed:
[추가] ProductController에 CrossOrigin 추가

jangsc0000 committed:
[리팩토링] 리터럴 static final 변수로 변경
- NOT_EXISTED_USER_ID = 0L

jangsc0000 committed:
[리팩토링] Exception message를 String.format으로 변경

jangsc0000 committed:
[리팩토링] 줄변경

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,98 @@
+     *
+     * @param id 유저 id
+     * @return 주어진 id와 일치하는 유저
+     * @throws UserNotFoundException '유저를 찾지 못했다'는 예외
```
`@throws`에 대해서는 "~한 경우" 처럼 작성해 보세요.
```suggestion
     * @throws UserNotFoundException 유저를 찾지 못한 경우
```

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/UserController.java:
```diff
@@ -0,0 +1,68 @@
+     *
+     * @param id   요청 id
+     * @param user 요청 유저
+     * @return
```
여기에 설명이 빠졌군요.

johngrib commented on app/src/main/java/com/codesoom/assignment/dto/UserData.java:
```diff
@@ -0,0 +1,20 @@
+import javax.validation.constraints.NotBlank;
+
+@Getter
+@Setter
```
`@Setter`를 제거해 보세요.

jangsc0000 replied:
Setter를 제거해야하는 이유를 찾아보고 있었는데.. 이펙티브 자바에서 찾았습니다!

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,239 @@
+            @Test
+            @DisplayName("'유저를 찾을수 없다'는 예외를 던진다.")
+            void it_throws_not_found_user_exception() {
+                assertThatThrownBy(() -> subject(NOT_EXISTED_USER_ID, updateUser))
```
`NOT_EXISTED_USER_ID`가 존재하지 않는 유저의 아이디라는 것을 어떻게 보장하나요? 지금은 그냥 0 이죠? 코드를 작성한 승철님은 0 은 괜찮다고 생각할 수 있습니다. 코드를 알고 있으니까요. 그런데 지금처럼 프로젝트가 소규모가 아니라 몇백만줄의 코드가 쌓여있는 거대한 프로젝트가 되었다고 생각해보세요. `0`이 정말 괜찮은지 아닌지 알아내려면 클래스 수십개를 열어봐야 할 수 있습니다. `NOT_EXISTED_USER_ID`에 논리적인 근거를 마련해 주세요. 저라면 `Context`에서 저 아이디를 찾아 삭제해주는 `@BeforeEach`를 추가해 줄 것 같군요.

jangsc0000 replied:
컨텍스트 내에서 논리적인 근거를 마련하라.. 명심하겠습니다

jangsc0000 committed:
[리팩토링] Setter 제거

jangsc0000 committed:
[리팩토링] Setter 테스트 제거

jangsc0000 committed:
[추가] Dozer Mapping 추가

jangsc0000 committed:
[리팩토링] 컨텍스트 근거 추가

johngrib commented on app/build.gradle:
```diff
@@ -31,6 +33,9 @@ repositories {
 
 dependencies {
+    // Dozer Mapper
+    implementation 'com.github.dozermapper:dozer-core:6.4.0'
```
`dozermapper`같은 매퍼는 실제 회사에서 잘 사용하지 않으며, 오히려 꺼려하는 경우도 많습니다. 그래서 이렇게 새로운 라이브러리를 추가할 때에는 제가 지금 코드 리뷰를 남기는 것처럼 본인의 PR에 코드 리뷰를 남겨서 이 라이브러리를 왜 추가했는지 설명을 해주시면 좋습니다. 동료를 설득하기 위해서요.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,98 @@
+     * @return 생성한 유저
+     */
+    public User createUser(UserData userData) {
+        User user = mapper.map(userData, User.class);
```
mapper가 도입된 상태에서 만약 mapper를 들어내야 하는 상황이 되었다고 생각해 보세요. 클래스가 100개가 있는데 클래스 100개에서 모두 mapper를 쓰고 있습니다. 어떻게 작업해야 할까요?

johngrib commented on app/src/test/java/com/codesoom/assignment/domain/UserTest.java:
```diff
@@ -0,0 +1,31 @@
+        void it_returns_user() {
+            User user = User.builder()
+                    .id(1L)
+                    .email("honggd@gmail.com")
+                    .name("홍길동")
+                    .password("password")
+                    .build();
+
+            assertThat(user.getId()).isEqualTo(1L);
+            assertThat(user.getEmail()).isEqualTo("honggd@gmail.com");
+            assertThat(user.getName()).isEqualTo("홍길동");
+            assertThat(user.getPassword()).isEqualTo("password");
```
오타로 인한 테스트 실패를 방지할 수 있게 중복되는 값들은 상수로 선언해 주세요.

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -0,0 +1,233 @@
+            }
+        }
+
+        @Nested
+        @DisplayName("등록되지 않은 유저의 id가 주어진다면")
+        class Context_with_not_existed_user_id {
+            User notExistedUser;
+
+            @BeforeEach
+            void prepare() {
+                prepareExistedUser();
+                userRepository.delete(existedUser);
+                notExistedUser = existedUser;
+            }
+
+            @Test
+            @DisplayName("Not Found를 응답한다.")
+            void it_responses_not_found() throws Exception {
+                mockMvc.perform(delete("/users/" + notExistedUser.getId()))
+                        .andExpect(status().isNotFound());
+            }
+        }
```
테스트를 아주 잘 작성하셨습니다.

johngrib reviewed (commented):
한 주간 고생 많으셨습니다. 주간 회고를 작성해 공유해주세요.
